src/allmydata/testing/web.py

```diff
@@ -0,0 +1,225 @@
+import hashlib
+
+import attr
+
+from hyperlink import DecodedURL
+
+from twisted.web.resource import (
+    Resource,
+)
+from twisted.web.iweb import (
+    IBodyProducer,
+)
+from twisted.internet.defer import (
+    succeed,
+)
+
+from treq.client import (
+    HTTPClient,
+    FileBodyProducer,
+)
+from treq.testing import (
+    RequestTraversalAgent,
+)
+from zope.interface import implementer
+
+import allmydata.uri
+from allmydata.util import (
+    base32,
+)
+
+
+class _FakeTahoeRoot(Resource, object):
+    """
+    This is a sketch of how an in-memory 'fake' of a Tahoe
+    WebUI. Ultimately, this will live in Tahoe
+    """
+
+    def __init__(self, uri=None):
+        Resource.__init__(self)  # this is an old-style class :(
+        self._uri = uri
+        self.putChild(b"uri", self._uri)
+
+    def add_data(self, kind, data):
+        return self._uri.add_data(kind, data)
+
+
+@attr.s
+class _FakeCapability(object):
+    """
+    """
+    data=attr.ib()
+
+
+KNOWN_CAPABILITIES = [
+    getattr(allmydata.uri, t).BASE_STRING
+    for t in dir(allmydata.uri)
+    if hasattr(getattr(allmydata.uri, t), 'BASE_STRING')
```

Comment:
`hasattr` used to be considered unsafe because it swallowed exceptions ... but maybe that's history, now?  It doesn't appear to be the case on Python 3.7 at least.